Mailbox 266 Update to ElasticSearch 2.2.1

chibenwa opened the pull request:
In this PR : 
 - I restore a test that was deleted
 - I update Guava to 18.0 as it is needed by ElasticSearch 2.2.1
 - I reworked the query translation (no more filteredQueries but instead BoolQuery with a filter argument)
 - I implemented delete by query in a htread pool on our side
 - I solved some other monir API changed
 - Cleaned a little POMs

That being done, I also reworked the way the indexer worked 

Why ?

Because using a delete with MessageRange::from I would make James iterate over ~2^64 uid, crashing a thread.
Because a user could well specify a range arbitrary long (2 to 2^63-1 for example) and cause the same symptomes
Because an update on 100 different mails would have resulted in 100 requests. I bulked that.

Ah and of course CI will not succeed to deploy that => wrong version of ElasticSearch !

chibenwa commented:
https://github.com/linagora/james-jenkins/pull/41 is needed as we need to run integration tests aginst a 2.2.1 ElasticSearch

aduprat commented on server/container/guice/cassandra-guice/pom.xml:
```diff
@@ -265,6 +265,10 @@
+                <dependency>
+                    <groupId>org.mockito</groupId>
+                    <artifactId>mockito-core</artifactId>
+                </dependency>
```
no scope ?

chibenwa replied:
Corrected

aduprat commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.msgpack.core.Preconditions;
```
com.google.common.base

chibenwa replied:
Corrected

aduprat commented on mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java:
```diff
@@ -129,6 +130,12 @@ public long getUidTo() {
     }
 
+    public long size() {
+        if (type == Type.FROM || type == Type.ALL) {
```
`Type.FROM` as a size of 1, no?

chibenwa replied:
Nope. Type FROM indicates that this MessageRange contains all uids of the mailbox superior to from date.
Thus, we can not know the size of this message range.
By the way, after reading matthieu comments, I realized they were no need to hndle deletes using MessageRange if I reworl ListeningSearchIndex class a little. I deleted this changeset and will propose a new one, of higher quality.

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/IndexCreationFactory.java:
```diff
@@ -69,22 +66,22 @@ private static ClientProvider createIndex(ClientProvider clientProvider, XConten
-            .field("number_of_replicas", nbReplica);
+    private static Settings generateSetting(int nbShards, int nbReplica, Optional<Settings> store) throws IOException {
+        Settings.Builder settingsBuilder = Settings.builder();
+        settingsBuilder.put("number_of_shards", nbShards);
```
you are supposed to chain calls on builders

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java:
```diff
@@ -103,126 +106,135 @@ private void registerHeaderOperatorConverters() {
+    private QueryBuilder convertConjunction(SearchQuery.ConjunctionCriterion criterion) {
+        switch (criterion.getType()) {
+            case AND:
+                return convertToBoolQuery(criterion.getCriteria(), BoolQueryBuilder::must);
```
it looks like you only want to convert type to BoolQueryBuilder

mbaechler commented on mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java:
```diff
@@ -60,8 +62,12 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(ElasticSearchIntegrationTest.class);
 
+    private TemporaryFolder temporaryFolder = new TemporaryFolder();
```
why did you kept `settingForInMemory` method earlier in the code ?

chibenwa replied:
This setting is still supported using the support client, but no more supported by the Embedded ES.

I deleted these lines in the next version of the code.

mbaechler commented on mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/query/QueryConverterTest.java:
```diff
@@ -50,14 +50,14 @@ public void allCriterionShouldBeWellConverted()throws Exception{
         assertThatJson(queryConverter.from(searchQuery, MAILBOX_UUID).toXContent(jsonBuilder(), QueryBuilder.EMPTY_PARAMS).string())
```
what is the value of such tests ?

chibenwa replied:
Close to 0 :
Have a place in the code where we actually see the query we are issuing, I guess.
In the next version of the code, I deleted them.

mbaechler replied:
Agree.
It's very important to write such tests while doing TDD but once everything is in place, you can remove the tests.
It's an usual TDD step.

mbaechler commented on mpt/impl/imap-mailbox/elasticsearch/src/test/java/org/apache/james/mpt/imapmailbox/elasticsearch/host/ElasticSearchHostSystem.java:
```diff
@@ -82,12 +76,32 @@ public boolean addUser(String user, String password) throws Exception {
     protected void resetData() throws Exception {
-        embeddedElasticSearch.after();
-        this.embeddedElasticSearch = new EmbeddedElasticSearch();
+        after();
```
it doesn't really make sense to me

mbaechler commented on mpt/impl/imap-mailbox/elasticsearch/src/test/java/org/apache/james/mpt/imapmailbox/elasticsearch/host/ElasticSearchHostSystem.java:
```diff
@@ -82,12 +76,32 @@ public boolean addUser(String user, String password) throws Exception {
+    }
+
+    private void before() throws Throwable {
+        this.publicTemporaryFolder = new PublicTemporaryFolder();
```
what is this PublicTemporaryFolder thing ?

mbaechler commented on mpt/impl/imap-mailbox/elasticsearch/src/test/java/org/apache/james/mpt/imapmailbox/elasticsearch/host/ElasticSearchHostSystem.java:
```diff
@@ -82,12 +76,32 @@ public boolean addUser(String user, String password) throws Exception {
+    private void before() throws Throwable {
+        this.publicTemporaryFolder = new PublicTemporaryFolder();
+        this.embeddedElasticSearch = new EmbeddedElasticSearch(publicTemporaryFolder);
+        publicTemporaryFolder.before();
```
you should call `before` before giving the object to EmbeddedElasticSearch

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+        executor.shutdownNow();
+    }
+
+    public void deleteByQuery(QueryBuilder queryBuilder) {
```
no return value ?

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+        try (Client client = clientProvider.get()) {
+            ListenableActionFuture<SearchResponse> searchResponseFuture = client.prepareSearch(ElasticSearchIndexer.MAILBOX_INDEX)
+                .setTypes(ElasticSearchIndexer.MESSAGE_TYPE)
+                .setScroll(new TimeValue(60000))
```
magic number ?

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+                if (searchResponse.getHits().getHits().length == 0) {
+                    break;
+                }
+                searchResponseFuture = client.prepareSearchScroll(searchResponse.getScrollId())
```
variable reuse ? really ?

chibenwa replied:
This code was taken from https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/java-search-scrolling.html

I reworked it to use future. But sure, I will completly rework it. And I should have done that first.

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+
+            while(true) {
+                SearchResponse searchResponse = searchResponseFuture.actionGet();
+                if (searchResponse.getHits().getHits().length == 0) {
```
give this condition a name, please

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+                .setSize(batchSize)
+                .execute();
+
+            while(true) {
```
I really don't like that : I would rather create a supplier that would supply some SearchResponse for each loop turn, or anything that would be higher level that what you did

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Inject
+    public DeleteByQueryPerformer(ClientProvider clientProvider,
+                                  @Named(DELETE_BY_QUERY_POOL_SIZE) Integer threadPoolSize,
```
I don't think we should have a threadPool by type of task.
What about creating a global unbounded thread pool for async tasks ?

chibenwa replied:
Sure, I will.

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexer.java:
```diff
@@ -65,8 +67,8 @@ public DeleteResponse deleteMessage(String id) {
-    public void deleteAllWithIdStarting(String idStart) {
-        throw new NotImplementedException();
+    public void deleteAllMatchingQuery(QueryBuilder queryBuilder) {
+        deleteByQueryPerformer.deleteByQuery(queryBuilder);
```
should by `.perform()` to be consistent

mbaechler commented on mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexerTest.java:
```diff
@@ -113,16 +125,15 @@ public void updateMessageShouldThrowWhenJsonIsNull() throws InterruptedException
 
-    @Ignore
     @Test
     public void deleteAllWithIdStarting() throws Exception {
```
method name doesn't sound good

mbaechler commented on mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexerTest.java:
```diff
@@ -134,28 +145,27 @@ public void deleteAllWithIdStarting() throws Exception {
+        String content = "{\"message\": \"trying out Elasticsearch\", \"mailboxId\":\"1\"}";
         
         testee.indexMessage(messageId, content);
         embeddedElasticSearch.awaitForElasticSearch();
```
why do you want at each step ?

mbaechler commented on dockerfiles/run/guice/destination/conf/elasticsearch.properties:
```diff
@@ -26,3 +26,8 @@ elasticsearch.nb.shards=1
+# This properties defines the maximal length of search results. Higher values
+# means more results but costs more performances.
+# Default value is 100
+elasticsearch.search.size=100
```
and so what ? we want to limit the search result for an imap query ?

chibenwa replied:
Ok, let speak a bit of ElasticSearch
See for instance https://www.elastic.co/guide/en/elasticsearch/reference/current/search-uri-request.html
You might find the size parament astounding : "The number of hits to return. Defaults to 10."
But... Why do ElasticSearch do that ?
Well, because of scalability concerns. Every shards returns its hits to the node performing the request. This node have to sort this list by relevance. As people are mostly concerned in general by search top results, a good optimisation is to limit the number of results returned (say 10) so that each shard can return its top 10 results, that will get ordere and trouncated on the node performing the query.
You have to give a size to your searches.
One workaround is using "Scrolls". Scrolls allow you to take back were you stopped.

For instance, with a query size of 10 : 
 
 -  step 1 get 10 first result of each shards and merge them : take 0->10
 - step 2 get 20 first results of each shards and merge them : take 10->20

 and so on.

By the way, in the first lines of the scroll API you have something that sounds not good : Scrolling is not intended for real time user requests, but rather for processing large amounts of data, e.g. in order to reindex the contents of one index into a new index with a different configuration.

We didn't implemented search using scroll API by lack of s/time/will/ . I remember having said it.

Well, this is for limiting response size of search in IMAP. To sum up, rapid implementation + technical limitation of ElasticSearch

Then for exposing to the configuration : 
 - I consider the more configuration you expose to the ADMIN, the better. Here he can trade accuracy vs performances.

I understand we disagree on the need to expose as many stuff as possible to configuration (of course with decent messages and a default value to back it up).

So I removed this change set.

But the size limitation will remain, with a constant value setted in some Java file.

chibenwa replied:
Oh I just ask the question : maybe you want me to implement this using a scroll ?

mbaechler replied:
Thank you very much for the explanation !

What bothers me is that I won't be able to found the mail I want when I need it.
From a business point of view, it's not ok to not be able to find a given email, even if there are much results.
So if we need scroll, let's do it.

chibenwa replied:
Ok, I'll add a commit for that on monday

mbaechler commented on mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java:
```diff
@@ -46,6 +46,7 @@
     public static final long NOT_A_UID = -1;
 
     public static final long MAX_UID = Long.MAX_VALUE;
+    public static final long UNKNOWN_SIZE = -1;
```
sorry ? unknown == -1 ? it's 2016 !

chibenwa replied:
After reading your comments, I realized they were no need to handle deletes using MessageRange if I rework ListeningSearchIndex class a little. I deleted this changeset and will propose a new one, of higher quality.

mbaechler commented on dockerfiles/run/guice/destination/conf/elasticsearch.properties:
```diff
@@ -41,4 +41,9 @@ elasticsearch.delete.by.query.batch.size=100
+# On messages ranges, if the range is bigger, JAMES will rely on a delete by query
+# instead of a bulk containing all the messages concerned
+# Default value is 100
+elasticsearch.maximum.delete.size.before.delete.by.query
```
what is the goal of this feature ?

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:
```diff
@@ -84,22 +94,63 @@ public void add(MailboxSession session, Mailbox<Id> mailbox, MailboxMessage<Id>
-                    }
-                });
+            } else {
+                indexer.deleteMessages(StreamSupport.stream(Spliterators.spliteratorUnknownSize(range.iterator(), Spliterator.ORDERED), true)
```
please, it's really really ugly, you are not supposed to propose such code

chibenwa replied:
A nicer API from ListeningMessageSearchIndex will remove the need from handling MessageRange as an iterator and will remove this uggly line.

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:
```diff
@@ -84,22 +94,63 @@ public void add(MailboxSession session, Mailbox<Id> mailbox, MailboxMessage<Id>
+        }
+    }
+
+    private void deleteSingleMessage(Mailbox<Id> mailbox, MessageRange range) {
```
really bad abstraction : what if range has several elements ?

chibenwa replied:
Changing ListeningMessageSearch index API to something better suited will remove the need of deleting individual messages.

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexer.java:
```diff
@@ -26,14 +26,34 @@
 
 public class ElasticSearchIndexer {
 
+    public static class UpdatedDocumentPart {
```
I don't understand what it means

chibenwa replied:
Its the representation of an update that the indexer should performed.

I just updated this inner class name

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexer.java:
```diff
@@ -26,14 +26,34 @@
+            return id;
+        }
+
+        public String getUpdatedDocumentPart() {
```
UpdatedDocumentPart.getUpdatedDocumentPart ?

chibenwa replied:
Having updated the class name this field makes more sens

mbaechler commented:
do you think there is enough high level tests to exercise your modifications ?
Maybe running them with a code coverage tool would be a start ...

Also, your branch has 61 warnings under eclipse.

chibenwa commented:
Ok...
On the code coverage tool : 
 - ElasticSearchIndexer : 100% for methods and lines
 - DeleteByQueryPerformer : 100% for methods and lines
 - ElasticSearchListeningMessageSearchIndex : 100% for methods, 95% for lines. The non convered lines are for propagating JsonProcessingException. Ok my bad I'll add a test for this edge case.
On query convesion the tests are slightly weaker, but I did not change those tests, so I do not feel guilty on this.

That being said, with the changes I will propose for deletes (similar to those seen here for updates) the idea will be to not relly any more on MessageRange (a complicated and inapropriate data representation) but rather on some more simple data and build the bulk requests from that).

Well, if you want, I will add more data on tests for the indexer. And also play more MPT tests on ES (deletion + updates on messages ranges) to show I have no regression in the test I am doing.

And for the Eclpse warning : 
 - what does "you introduced 61 warning" means exactly ? Is it "in the files you modified, you have 61 warnings" or "your change sets introduced 61 warnings" ? There is a difference... Well, I will try to comply on this and will install eclipse to see that by myself and correct it.

mbaechler commented:
very cool about the coverage results ! Do what you think is needed to have the right coverage, I trust you.

For the warnings : master doesn't contain any warning under eclipse (it's a kind of a rule), so when I say the branch contains 61 warnings, it's 61 warning you introduced.

chibenwa commented:
MPT tests already covers updates.
I introduced a Scroll Iterables and use it for search and delete by query. I think the code is more readable...
I also corrected a concurrency issue in tests.
I introduced a coomon thread pool for async tasks. Not putting its name in a constant was done on purpose (avoids dependency).
I also bulked uodates and deletes. Changing the API of ListeningMessageSearchIndex makes it more readable, in my opinion.

aduprat commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ScrollIterable.java:
```diff
@@ -0,0 +1,80 @@
+        public SearchResponse next() {
+            SearchResponse result = searchResponseFuture.actionGet();
+            searchResponseFuture =  client.prepareSearchScroll(result.getScrollId())
+                .setScroll(new TimeValue(60000))
```
Magic value

aduprat commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,81 @@
+            for (SearchResponse searchResponse : scrollIterable) {
+                deleteRetrievedIds(client, searchResponse);
+            }
+            return null;
```
Not needed

aduprat commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -0,0 +1,81 @@
+
+    private ListenableActionFuture<BulkResponse> deleteRetrievedIds(Client client, SearchResponse searchResponse) {
+        BulkRequestBuilder bulkRequestBuilder = client.prepareBulk();
+        for (SearchHit hit : searchResponse.getHits()) {
```
stream.forEach

aduprat commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -35,16 +35,22 @@
+        this(clientProvider, executor, DEFAULT_BATCH_SIZE);
+    }
+
+    public DeleteByQueryPerformer(ClientProvider clientProvider, @Named("AsyncExecutor") ExecutorService executor, int batchSize) {
```
@VisibleForTesting

aduprat commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java:
```diff
@@ -41,20 +41,26 @@
+        this.isSynchronous = isSynchronous;
     }
 
     public void perform(QueryBuilder queryBuilder) {
```
Doesn't it make sense to override a method in tests than to introduce some code used only for testing in your implementation?

aduprat commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AbstractMessageMapperTest.java:
```diff
@@ -444,7 +444,7 @@ public void copyShouldCreateAMessageInDestination() throws MailboxException, IOE
         assertThat(messageMapper.getLastUid(benwaInboxMailbox)).isGreaterThan(message6.getUid());
 
-        MailboxMessage result = messageMapper.findInMailbox(benwaInboxMailbox,
+        MailboxMessage<Id> result = messageMapper.findInMailbox(benwaInboxMailbox,
```
Already on master

aduprat commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/SimpleMailboxMessageTest.java:
```diff
@@ -60,15 +60,23 @@ public void testSize() {
         ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-        byteArrayOutputStream.write(MESSAGE.getFullContent());
-        assertThat(byteArrayOutputStream.size()).isEqualTo(MESSAGE_CONTENT.getBytes().length);
+        try {
```
Already on master

mbaechler commented on mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ScrollIterable.java:
```diff
@@ -0,0 +1,80 @@
+        return new ScrollIterator(client, searchRequestBuilder);
+    }
+
+    public Stream<SearchResponse> toStream() {
```
s/toStream/stream

aduprat commented:
test this please

chibenwa commented:
test this please

aduprat commented on server/data/data-library/src/main/java/org/apache/james/domainlist/lib/AbstractDomainList.java:
```diff
@@ -69,8 +69,12 @@ public void configure(HierarchicalConfiguration config) throws ConfigurationExce
     @Override
-    public String getDefaultDomain() {
-        return defaultDomain;
+    public String getDefaultDomain() throws DomainListException {
```
There's still a problem, you are replacing the NPE by another Exception; but the init will still fail, no?

chibenwa replied:
Nope.
I'm saying if it's null we had a problem with configuration. Rather than returning a null result, let's say to the world we had a problem.
And hopefully RemoteDelivery handles DomainList exception and already backs up to "localhost" if a DomainList exception is encountered. It means sender threads do not die stupidly anymore.
I think solving the initialization problem fully and cleanly is out of the scope of this PR and should require some more work.
This temporary work around makes sens to me, and solves the issue temporary.
And no it do not fail anymore.

aduprat replied:
Ok

aduprat commented:
test this please

aduprat commented:
:+1:

aduprat commented on mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMailboxMessageSearchIndexTest.java:
```diff
@@ -237,8 +238,8 @@ public void deleteAllShouldWork() throws Exception {
-        expect(indexer.deleteAllMatchingQuery(anyObject(QueryBuilder.class)))
-            .andReturn(null);
+        indexer.deleteAllMatchingQuery(anyObject(QueryBuilder.class));
+        EasyMock.expectLastCall();
```
import static
